File: win-prog/ProgWinVB/Edit List and Spin/Transform/MatrixMethods.vb
'----------------------------------------------
' MatrixMethods.vb (c) 2002 by Charles Petzold
'----------------------------------------------
Imports System
Imports System.Drawing
Imports System.Drawing.Drawing2D
Imports System.Windows.Forms

Class MatrixMethods
    Inherits Form

    Private matx As Matrix
    Private btnInvert As Button
    Private updown(2) As NumericUpDown
    Private radbtn(1) As RadioButton

    Sub New()
        Text = "Matrix Methods"
        FormBorderStyle = FormBorderStyle.FixedDialog
        ControlBox = False
        MinimizeBox = False
        MaximizeBox = False
        ShowInTaskbar = False
        Location = Point.op_Addition(ActiveForm.Location, _
                   Size.op_Addition(SystemInformation.CaptionButtonSize, _
                                    SystemInformation.FrameBorderSize))

        Dim astrLabel() As String = {"X / DX:", "Y / DY:", "Angle:"}
        Dim i As Integer

        For i = 0 To 2
            Dim lbl As New Label()
            lbl.Parent = Me
            lbl.Text = astrLabel(i)
            lbl.Location = New Point(8, 8 + 16 * i)
            lbl.Size = New Size(32, 8)

            updown(i) = New NumericUpDown()
            updown(i).Parent = Me
            updown(i).Location = New Point(40, 8 + 16 * i)
            updown(i).Size = New Size(48, 12)
            updown(i).TextAlign = HorizontalAlignment.Right
            updown(i).DecimalPlaces = 2
            updown(i).Increment = 0.1D
            updown(i).Minimum = Decimal.MinValue
            updown(i).Maximum = Decimal.MaxValue
        Next i

        ' Create group box and radio buttons.
        Dim grpbox As New GroupBox()
        grpbox.Parent = Me
        grpbox.Text = "Order"
        grpbox.Location = New Point(8, 60)
        grpbox.Size = New Size(80, 32)

        For i = 0 To 1
            radbtn(i) = New RadioButton()
            radbtn(i).Parent = grpbox
            radbtn(i).Text = New String() {"Prepend", "Append"}(i)
            radbtn(i).Location = New Point(8, 8 + 12 * i)
            radbtn(i).Size = New Size(50, 10)
            radbtn(i).Checked = (i = 0)
        Next i

        ' Create 8 buttons for terminating dialog.
        Dim astrButton() As String = {"Reset", "Invert", "Translate", _
                                      "Scale", "Rotate", "RotateAt", _
                                      "Shear", "Cancel"}
        Dim aeh() As EventHandler = {AddressOf ButtonResetOnClick, _
                                     AddressOf ButtonInvertOnClick, _
                                     AddressOf ButtonTranslateOnClick, _
                                     AddressOf ButtonScaleOnClick, _
                                     AddressOf ButtonRotateOnClick, _
                                     AddressOf ButtonRotateAtOnClick, _
                                     AddressOf ButtonShearOnClick}
        For i = 0 To 7
            Dim btn As New Button()
            btn.Parent = Me
            btn.Text = astrButton(i)
            btn.Location = New Point(100 + 72 * (i \ 4),8 + 24 * (i Mod 4))
            btn.Size = New Size(64, 14)

            If i = 0 Then AcceptButton = btn
            If i = 1 Then btnInvert = btn

            If i < 7 Then
                AddHandler btn.Click, aeh(i)
                btn.DialogResult = DialogResult.OK
            Else
                btn.DialogResult = DialogResult.Cancel
                CancelButton = btn
            End If
        Next i

        ClientSize = New Size(240, 106)
        AutoScaleBaseSize = New Size(4, 8)
    End Sub

    Property Matrix() As Matrix
        Set(ByVal Value As Matrix)
            matx = Value
            btnInvert.Enabled = Matrix.IsInvertible
        End Set
        Get
            Return matx
        End Get
    End Property

    Private Sub ButtonResetOnClick(ByVal obj As Object, _
                                   ByVal ea As EventArgs)
        Matrix.Reset()
    End Sub

    Private Sub ButtonInvertOnClick(ByVal obj As Object, _
                                    ByVal ea As EventArgs)
        Matrix.Invert()
    End Sub

    Private Sub ButtonTranslateOnClick(ByVal obj As Object, _
                                       ByVal ea As EventArgs)
        Matrix.Translate(CSng(updown(0).Value), _
                         CSng(updown(1).Value), PrependOrAppend)
    End Sub

    Private Sub ButtonScaleOnClick(ByVal obj As Object, _
                                   ByVal ea As EventArgs)
        Matrix.Scale(CSng(updown(0).Value), _
                     CSng(updown(1).Value), PrependOrAppend)
    End Sub

    Private Sub ButtonRotateOnClick(ByVal obj As Object, _
                                    ByVal ea As EventArgs)
        Matrix.Rotate(CSng(updown(2).Value), PrependOrAppend)
    End Sub

    Private Sub ButtonRotateAtOnClick(ByVal obj As Object, _
                                      ByVal ea As EventArgs)
        Matrix.RotateAt(CSng(updown(2).Value), _
                        New PointF(CSng(updown(0).Value), _
                                   CSng(updown(1).Value)), PrependOrAppend)
    End Sub

    Private Sub ButtonShearOnClick(ByVal obj As Object, _
                                   ByVal ea As EventArgs)
        Matrix.Shear(CSng(updown(0).Value), _
                     CSng(updown(1).Value), PrependOrAppend)
    End Sub

    Private Function PrependOrAppend() As MatrixOrder
        If radbtn(0).Checked Then
            Return MatrixOrder.Prepend
        Else
            Return MatrixOrder.Append
        End If
    End Function
End Class

File: win-prog/ProgWinCS/Edit List and Spin/Transform/MatrixMethods.cs
//--------------------------------------------
// MatrixMethods.cs © 2001 by Charles Petzold
//--------------------------------------------
using System;
using System.Drawing;
using System.Drawing.Drawing2D;
using System.Windows.Forms;

class MatrixMethods: Form
{
     Matrix          matrix; 
     Button          btnInvert;
     NumericUpDown[] updown = new NumericUpDown[3];
     RadioButton[]   radio = new RadioButton[2];

     public MatrixMethods() 
     {
          Text = "Matrix Methods";
          FormBorderStyle = FormBorderStyle.FixedDialog;
          ControlBox      = false;
          MinimizeBox     = false;
          MaximizeBox     = false;
          ShowInTaskbar   = false;
          Location        = ActiveForm.Location +
                            SystemInformation.CaptionButtonSize +
                            SystemInformation.FrameBorderSize;

          String[] astrLabel = { "X / DX:", "Y / DY:", "Angle:" };

          for (int i = 0; i < 3; i++)
          {
               Label label    = new Label();
               label.Parent   = this;
               label.Text     = astrLabel[i];
               label.Location = new Point(8, 8 + 16 * i);
               label.Size     = new Size(32, 8);

               updown[i] = new NumericUpDown();
               updown[i].Parent        = this;
               updown[i].Location      = new Point(40, 8 + 16 * i);
               updown[i].Size          = new Size(48, 12);
               updown[i].TextAlign     = HorizontalAlignment.Right;

               updown[i].DecimalPlaces = 2;
               updown[i].Increment = 0.1m;
               updown[i].Minimum = Decimal.MinValue;
               updown[i].Maximum = Decimal.MaxValue;
          }
               // Create group box and radio buttons.

          GroupBox grpbox = new GroupBox();
          grpbox.Parent   = this;
          grpbox.Text     = "Order";
          grpbox.Location = new Point(8, 60);
          grpbox.Size     = new Size(80, 32);

          for (int i = 0; i < 2; i++)
          {
               radio[i] = new RadioButton();
               radio[i].Parent = grpbox;
               radio[i].Text = new string[] { "Prepend", "Append" } [i];
               radio[i].Location = new Point(8, 8 + 12 * i);
               radio[i].Size     = new Size(50, 10);
               radio[i].Checked = (i == 0);
          }

               // Create 8 buttons for terminating dialog.

          string[] astrButton = { "Reset", "Invert", "Translate", "Scale",
                                  "Rotate", "RotateAt", "Shear", "Cancel" };

          EventHandler[] aeh = { new EventHandler(ButtonResetOnClick), 
                                 new EventHandler(ButtonInvertOnClick),
                                 new EventHandler(ButtonTranslateOnClick), 
                                 new EventHandler(ButtonScaleOnClick),
                                 new EventHandler(ButtonRotateOnClick), 
                                 new EventHandler(ButtonRotateAtOnClick),
                                 new EventHandler(ButtonShearOnClick) };

          for (int i = 0; i < 8; i++)
          {
               Button btn = new Button();
               btn.Parent = this;
               btn.Text   = astrButton[i];
               btn.Location = new Point(100 + 72 * (i > 3 ? 1 : 0), 
                                        8 + (i % 4) * 24);
               btn.Size     = new Size(64, 14);
     
               if (i == 0)    // Reset button
               {
                    AcceptButton = btn;
               }
               if (i == 1)    // Invert button
               {
                    btnInvert = btn;
               }
               if (i < 7)     // All buttons except Cancel
               {
                    btn.Click += aeh[i];
                    btn.DialogResult = DialogResult.OK;
               }
               else           // Cancel button
               {
                    btn.DialogResult = DialogResult.Cancel;
                    CancelButton = btn;
               }
          }
          ClientSize = new Size(240, 106);

          AutoScaleBaseSize = new Size(4, 8);
     }
     public Matrix Matrix 
     {
          get 
          { 
               return matrix; 
          } 
          set 
          { 
               matrix = value;
               btnInvert.Enabled = matrix.IsInvertible;
          } 
     }
     void ButtonResetOnClick(object obj, EventArgs ea)
     {
          matrix.Reset();
     }
     void ButtonInvertOnClick(object obj, EventArgs ea)
     {
          matrix.Invert();
     }
     void ButtonTranslateOnClick(object obj, EventArgs ea)
     {
          matrix.Translate((float) updown[0].Value, 
                           (float) updown[1].Value,
               radio[0].Checked ? MatrixOrder.Prepend : MatrixOrder.Append);
     }
     void ButtonScaleOnClick(object obj, EventArgs ea)
     {
          matrix.Scale((float) updown[0].Value, 
                       (float) updown[1].Value,
               radio[0].Checked ? MatrixOrder.Prepend : MatrixOrder.Append);
     }
     void ButtonRotateOnClick(object obj, EventArgs ea)
     {
          matrix.Rotate((float) updown[2].Value,
               radio[0].Checked ? MatrixOrder.Prepend : MatrixOrder.Append);
     }
     void ButtonRotateAtOnClick(object obj, EventArgs ea)
     {
          matrix.RotateAt((float) updown[2].Value, 
                          new PointF((float) updown[0].Value, 
                                     (float) updown[1].Value),
               radio[0].Checked ? MatrixOrder.Prepend : MatrixOrder.Append);
     }
     void ButtonShearOnClick(object obj, EventArgs ea)
     {
          matrix.Shear((float) updown[0].Value, 
                       (float) updown[1].Value,
               radio[0].Checked ? MatrixOrder.Prepend : MatrixOrder.Append);
     }
}